### ISSUE TITLE
On a mac: The cursor position could not be read within a normal duration

### ISSUE THREAD

**Issue by stormasm** [nushell/reedline#132]
Error: Custom { kind: Other, error: "The cursor position could not be read within a normal duration" }

**Comment by jntrnr** [comment]
@stormasm do you know roughly how long it takes? (like at least an hour?)

**Comment by stormasm** [comment]
@jntrnr yes its a long time  --- I believe way more than one hour but less than one day.  I have all of these different reedline terminals up including now engine-q and nucli-reed so every once in awhile I see this message --- was wondering if anyone else has seen it ?  I am running a Mac Catalina 10.15.6 --- not that it matters but just an FYI.

**Comment by sholderbach** [comment]
Might be something we have to file with `crossterm` There seems to be another unrelated crash https://github.com/crossterm-rs/crossterm/issues/367

The cursor position is retrieved in crossterm by flushing an Escape code to stdout and then polling for the event (with some kind of locking mechanism underneath) with a 2 sec timeout. 

https://github.com/crossterm-rs/crossterm/blob/24d0f561e8b5501c18f8e7fd18a60be0624dad68/src/cursor/sys/unix.rs#L30-L54

If we don't leak some resources on the `reedline` side this might be a system pathology caught by crossterm. 
Would you mind running a `/usr/bin/time` equivalent for Mac with the reedline app to maybe catch how long it takes to crash or how much peak memory was consumed?

**Comment by stormasm** [comment]
@sholderbach thanks I will research further and report back to you when I have more info ---- it could be a weird bug in my code as I am doing some experimentation with engine-q...

**Comment by jntrnr** [comment]
No repro on Linux - let it run yesterday, slept the laptop, and woke it up. Neither was able to break. May be OS-specific

**Comment by mzanrosso** [comment]
> No repro on Linux - let it run yesterday, slept the laptop, and woke it up. Neither was able to break. May be OS-specific

I can confirm that even my MacOS 12 beta 3 is affected by this problem

**Comment by jntrnr** [comment]
I'm giving it a try on macOS here so see if I can repro. Would like to see if we can get this fixed before the release. Something I'm hoping to do fairly soon is to integrate reedline into nushell, but if we're having stability issues that's not going to work.

**Comment by sholderbach** [comment]
I also tried to provoke the issue under Linux in gnome-terminal using different starting conditions for the reedline demo app but was not able to catch an error. 
I ran `perf` on an idling demo app prompt, and with the `full_repaint` triggered for the refreshing clock, the `crossterm::cursor::position` function (throwing the Error) is basically part of the hotpath.

**Comment by sholderbach** [comment]
But even trying to poll the cursor position every millisecond (changing the Event polling in reedline) doesn't trigger the Error for me (even though the terminal emulator process starts to consume some serious CPU). Crossterm does not specify any contract we have to uphold or mention any compatibility notes. We probably should bring this to their attention.

**Comment by jntrnr** [comment]
Seems like a good take-away here is that if we're animating, we're increasing the change of hitting a platform-specific issue. We may want to make the animating prompt something you can disable so that it's only drawn when necessary. That doesn't fix the issue, but it would decrease the chance of hitting it.

@sholderbach - yeah I'm feeling more convinced it's a macOS-specific issue. Still trying to find a repro

**Comment by sholderbach** [comment]
On our side we might consider if we get by with dropping the error and reusing a cached position or setting to a desired one explicitly and then retrying reading.

**Comment by jntrnr** [comment]
@sholderbach yeah I like that idea. at the very least, be a little more flexible if we do encounter errors.

If we get in a rare state where the position changes but we can't get it, the user should be able to hit enter and get back to a clean prompt

**Comment by sholderbach** [comment]
This one seems to be also happening with an ssh session into Linux (WSL) https://github.com/crossterm-rs/crossterm/issues/367

**Comment by jntrnr** [comment]
Looks like we don't handle any failure in our demo. I've updated it here: https://github.com/jntrnr/reedline/commit/289188eee746eb742dc6c34ea7aa7028cee6def9

@mzanrosso / @stormasm - can you try again with the latest source? You'll still see an error if it happens, but hopefully it'll pick back up and give you a new prompt+input to work with.

**Comment by stormasm** [comment]
Ok I grabbed the latest reedline and engine-q and will let them both run overnight...

**Comment by mzanrosso** [comment]
> Looks like we don't handle any failure in our demo. I've updated it here: https://github.com/jntrnr/reedline/commit/289188eee746eb742dc6c34ea7aa7028cee6def9
> 
> 
> 
> @mzanrosso / @stormasm - can you try again with the latest source? You'll still see an error if it happens, but hopefully it'll pick back up and give you a new prompt+input to work with.

I will try asap, now I'm a bit busy with other stuff here; but thanks for the commit!

**Comment by stormasm** [comment]
Both engine-q and the reedline demo ran overnight and in the morning (of course) there is no error showing up on either one... So far, so good ---- will report here if / when I see something again.

**Comment by stormasm** [comment]
bingo ! just reproduced the error after firing up terminal in less than 5 minutes...

Error: Custom { kind: Other, error: "The cursor position could not be read within a normal duration" }

so it crashed on my Mac in less than 5 minutes running engine-q

**Comment by jntrnr** [comment]
@stormasm - oh we should fix engine-q also. The fix I did was to the demo for reedline, and I'm betting engine-q needs the same fix. I'll take a look

**Comment by jntrnr** [comment]
@stormasm - just checked in a similar fix to engine-q that I used for reedline

**Comment by stormasm** [comment]
@jntrnr ok sounds good !  I saw the fix you did to reedline, also you may want to put into the Cargo.toml file in engine-q to pull in the branch = main

```rust
reedline = {git = "https://github.com/jntrnr/reedline", branch = "main"}
````

**Comment by stormasm** [comment]
and possibly a .gitignore file too...

**Comment by stormasm** [comment]
ok cool ---- just saw your changes ---- thanks...

**Comment by stormasm** [comment]
Caused by:
  failed to load source for dependency `reedline`

Caused by:
  Unable to update https://github.com/jntrnr/reedline

Caused by:
  failed to find branch `master`

Caused by:
  cannot locate remote-tracking branch 'origin/master'; class=Reference (4); code=NotFound (-3)

**Comment by jntrnr** [comment]
@stormasm - I wonder what is causing it. `cargo update` seems to run okay for me.

**Comment by jntrnr** [comment]
```
> rustc --version
rustc 1.54.0-nightly (ed597e7e1 2021-06-08)
```

I can switch to stable and see if I can repro

**Comment by jntrnr** [comment]
Yeah stable does it. @stormasm if you want to submit the branch PR, I can land it

**Comment by jntrnr** [comment]
@stormasm - went ahead and landed your fix

**Comment by jntrnr** [comment]
Any repros after the fixes landed?

**Comment by stormasm** [comment]
so far I have not seen any issues, but I have been away for a few days and have not banged on it too much...  I will hopefully be spending more time this week with engine-q / and that is where I saw the problem last --- will report back here if I can reproduce again this week...

**Comment by stormasm** [comment]
@jntrnr this error might be relevant or helpful...

```
Error: Custom { kind: Other, error: "The cursor position could not be read within a normal duration" }
```

**Comment by jntrnr** [comment]
@stormasm seems we're getting that as the same error. That error, I believe, is coming from crossterm since it's not coming from reedline.

We could try reporting it to them. Maybe they've seen it before?

**Comment by stormasm** [comment]
Error is coming from 
[here in crossterm](https://github.com/crossterm-rs/crossterm/blob/master/src/cursor/sys/unix.rs#L48)
as you stated...

```
src$ rg 'position\(\)'
engine.rs
675:            let (column, row) = position()?;
743:                                    let (original_column, original_row) = position()?;
748:                                    let (new_column, _) = position()?;
799:                        prompt_origin.1 = position()?.1.saturating_sub(1);

painter.rs
165:        let prompt_offset = position()?;
src$ 
```

so basically coming from our method in the painter.rs 
```
    pub fn repaint_everything(
```

or possibly this method in the engine.rs
```
    fn read_line_helper(&mut self, prompt: &dyn Prompt) -> Result<Signal> {

```

**Comment by stormasm** [comment]
So I removed the timestamp from the prompt to see if that is what is causing the bug...

https://github.com/stormasm/reedline/tree/notime

But I still see the repaint happening every second when I put a println! inside the engine-q...

Rustyline doesn't work this way repainting like this because I never see this happening (I don't think) when I am debugging nushell...

Why are we doing it this way in reedline ?  

Instead of using an event based trigger to fire a repaint ?

**Comment by stormasm** [comment]
https://github.com/stormasm/reedline/tree/notime

I am doing some different experiments with this now; including removing the timestamp from the prompt...

**Comment by stormasm** [comment]
All of these comments are about developing with engine-q on my Mac.

So I removed the timestamp from the prompt...
https://github.com/stormasm/reedline/tree/notime
And the problem I stated above still persists randomly at times...
So drawing the timestamp every second is not the issue.

Also I am seeing some strange behavior with the Command-K key sequence.

If I hit Ctrl L it clears the screen as it should.
But if I hit a bunch of returns till I get in the middle of the screen,
and then hit Command-K it clears the screen like it does in bash,
but then redraws the cursor in the exact same spot it was at before it clears the screen.

So reedline is clearing the screen and then repainting the cursor in the exact same spot.
Very similar behavior as to what I see when I println! in a loop in engine-q
But this only happens when I am at the bottom of the screen.

If I Ctrl-L to the top --- then I do not see any repaint at all of the println! stuff...
